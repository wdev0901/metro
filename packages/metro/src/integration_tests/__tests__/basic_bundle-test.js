/**
 * Copyright (c) 2015-present, Facebook, Inc.
 * All rights reserved.
 *
 * This source code is licensed under the BSD-style license found in the
 * LICENSE file in the root directory of this source tree. An additional grant
 * of patent rights can be found in the PATENTS file in the same directory.
 *
 * @emails oncall+javascript_foundation
 * @format
 */

'use strict';

jest
  .useRealTimers()
  .unmock('fs')
  .unmock('graceful-fs');

const Metro = require('../..');
const path = require('path');

jasmine.DEFAULT_TIMEOUT_INTERVAL = 30 * 1000;

const INPUT_PATH = path.resolve(__dirname, '../basic_bundle');
const POLYFILLS_PATH = path.resolve(__dirname, '../../lib/polyfills');
const ASSET_REGISTRY_PATH = path.resolve(
  __dirname,
  '../basic_bundle/AssetRegistry',
);

describe('basic_bundle', () => {
  const absPathRe = new RegExp(INPUT_PATH, 'g');
  const polyfill1 = path.join(INPUT_PATH, 'polyfill-1.js');
  const polyfill2 = path.join(INPUT_PATH, 'polyfill-2.js');

  beforeEach(() => {
    jest.resetModules();

    // We replace the farm by a simple require, so that the worker sources are
    // transformed and managed by jest.
    jest.mock('jest-worker', () => {
      function Worker(workerPath, opts) {
        const {Readable} = require('stream');
        const worker = require(workerPath);
        const api = {
          getStdout: () => new Readable({read() {}}),
          getStderr: () => new Readable({read() {}}),
          end: () => (ended = true),
        };

        let ended = false;

        opts.exposedMethods.forEach(name => {
          api[name] = function() {
            if (ended) {
              throw new Error('worker farm was ended');
            }

            return worker[name].apply(null, arguments);
          };
        });

        return api;
      }

      return {
        __esModule: true,
        default: Worker,
      };
    });
  });

  afterEach(() => {
    jest.resetAllMocks();
  });

  /**
   * On different machines and local repos the absolute paths are different so
   * we relativize them to the root of the test. We also trim the spacing inside
   * ID-based `require()` calls because the way the bundle appears to do it is
   * by replacing path strings directly by numbers in the code, not the AST.
   */
  function verifyResultCode(code) {
    expect(
      code
        .replace(absPathRe, '')
        .replace(/require\(([0-9]*) *\)/g, 'require($1)'),
    ).toMatchSnapshot();
  }

  it('bundles package with polyfills', async () => {
    const bundleWithPolyfills = await Metro.build(
      {
        assetRegistryPath: ASSET_REGISTRY_PATH,
        dynamicDepsInPackages: 'reject',
        getModulesRunBeforeMainModule: () => ['InitializeCore'],
        getPolyfills: () => [polyfill1, polyfill2],
        projectRoots: [INPUT_PATH, POLYFILLS_PATH],
        transformCache: Metro.TransformCaching.none(),
        transformModulePath: require.resolve('../../transformer'),
        nonPersistent: true,
      },
      {
        dev: false,
        entryFile: path.join(INPUT_PATH, 'TestBundle.js'),
        platform: 'ios',
      },
    );
    verifyResultCode(bundleWithPolyfills.code);
  });

  it('bundles package without polyfills', async () => {
    const bundleWithoutPolyfills = await Metro.build(
      {
        assetRegistryPath: ASSET_REGISTRY_PATH,
        dynamicDepsInPackages: 'reject',
        getModulesRunBeforeMainModule: () => ['InitializeCore'],
        getPolyfills: () => [],
        projectRoots: [INPUT_PATH, POLYFILLS_PATH],
        transformCache: Metro.TransformCaching.none(),
        transformModulePath: require.resolve('../../transformer'),
        nonPersistent: true,
      },
      {
        dev: false,
        entryFile: path.join(INPUT_PATH, 'TestBundle.js'),
        platform: 'ios',
      },
    );
    verifyResultCode(bundleWithoutPolyfills.code);
  });
});
