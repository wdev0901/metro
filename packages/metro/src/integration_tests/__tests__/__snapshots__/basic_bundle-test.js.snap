// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`basic_bundle bundles package with polyfills 1`] = `
"var __DEV__=false,__BUNDLE_START_TIME__=this.nativePerformanceNow?nativePerformanceNow():Date.now(),process=this.process||{};process.env=process.env||{};process.env.NODE_ENV='production';
(function (global) {
  'use strict';

  global.require = _require;
  global.__d = define;
  var modules = Object.create(null);

  function define(factory, moduleId, dependencyMap) {
    if (moduleId in modules) {
      return;
    }

    modules[moduleId] = {
      dependencyMap: dependencyMap,
      exports: undefined,
      factory: factory,
      hasError: false,
      isInitialized: false
    };
  }

  function _require(moduleId) {
    var moduleIdReallyIsNumber = moduleId;
    var module = modules[moduleIdReallyIsNumber];
    return module && module.isInitialized ? module.exports : guardedLoadModule(moduleIdReallyIsNumber, module);
  }

  var inGuard = false;

  function guardedLoadModule(moduleId, module) {
    if (!inGuard && global.ErrorUtils) {
      inGuard = true;
      var returnValue = void 0;

      try {
        returnValue = loadModuleImplementation(moduleId, module);
      } catch (e) {
        global.ErrorUtils.reportFatalError(e);
      }

      inGuard = false;
      return returnValue;
    } else {
      return loadModuleImplementation(moduleId, module);
    }
  }

  var ID_MASK_SHIFT = 16;
  var LOCAL_ID_MASK = ~0 >>> ID_MASK_SHIFT;

  function unpackModuleId(moduleId) {
    var segmentId = moduleId >>> ID_MASK_SHIFT;
    var localId = moduleId & LOCAL_ID_MASK;
    return {
      segmentId: segmentId,
      localId: localId
    };
  }

  _require.unpackModuleId = unpackModuleId;

  function packModuleId(value) {
    return value.segmentId << ID_MASK_SHIFT + value.localId;
  }

  _require.packModuleId = packModuleId;

  function loadModuleImplementation(moduleId, module) {
    var nativeRequire = global.nativeRequire;

    if (!module && nativeRequire) {
      var _unpackModuleId = unpackModuleId(moduleId),
          _segmentId = _unpackModuleId.segmentId,
          _localId = _unpackModuleId.localId;

      nativeRequire(_localId, _segmentId);
      module = modules[moduleId];
    }

    if (!module) {
      throw unknownModuleError(moduleId);
    }

    if (module.hasError) {
      throw moduleThrewError(moduleId, module.error);
    }

    module.isInitialized = true;
    var exports = module.exports = {};
    var _module = module,
        factory = _module.factory,
        dependencyMap = _module.dependencyMap;

    try {
      var _moduleObject = {
        exports: exports
      };
      factory(global, _require, _moduleObject, exports, dependencyMap);
      {
        module.factory = undefined;
        module.dependencyMap = undefined;
      }
      return module.exports = _moduleObject.exports;
    } catch (e) {
      module.hasError = true;
      module.error = e;
      module.isInitialized = false;
      module.exports = undefined;
      throw e;
    }
  }

  function unknownModuleError(id) {
    var message = 'Requiring unknown module \\"' + id + '\\".';
    return Error(message);
  }

  function moduleThrewError(id, error) {
    var displayName = id;
    return Error('Requiring module \\"' + displayName + '\\", which threw an exception: ' + error);
  }
})(this);
(function (global) {
  'use strict';

  if (!Object.keys) {
    Object.keys = function () {};
  }
})(this);
(function (global) {
  'use strict';

  if (!String.prototype.repeat) {
    String.prototype.repeat = function () {};
  }
})(this);
__d(function (global, _require, module, exports, _dependencyMap) {
  'use strict';

  var Bar = _require(_dependencyMap[0]);

  var Foo = _require(_dependencyMap[1]);

  module.exports = {
    Foo: Foo,
    Bar: Bar
  };
},4,[5,6]);
__d(function (global, _require, module, exports, _dependencyMap) {
  'use strict';

  var Foo = _require(_dependencyMap[0]);

  module.exports = {
    type: 'bar',
    foo: Foo.type
  };
},5,[6]);
__d(function (global, _require, module, exports, _dependencyMap) {
  'use strict';

  var asset = _require(_dependencyMap[0]);

  module.exports = {
    type: 'foo',
    asset: asset
  };
},6,[7]);
__d(function (global, _require, module, exports, _dependencyMap) {
  module.exports = _require(_dependencyMap[0]).registerAsset({
    \\"__packager_asset\\": true,
    \\"httpServerLocation\\": \\"/assets\\",
    \\"width\\": 8,
    \\"height\\": 8,
    \\"scales\\": [1],
    \\"hash\\": \\"77d45c1f7fa73c0f6c444a830dc42f67\\",
    \\"name\\": \\"test\\",
    \\"type\\": \\"png\\"
  });
},7,[8]);
__d(function (global, _require, module, exports, _dependencyMap) {
  'use strict';
},8,[]);
require(4);"
`;

exports[`basic_bundle bundles package without polyfills 1`] = `
"var __DEV__=false,__BUNDLE_START_TIME__=this.nativePerformanceNow?nativePerformanceNow():Date.now(),process=this.process||{};process.env=process.env||{};process.env.NODE_ENV='production';
(function (global) {
  'use strict';

  global.require = _require;
  global.__d = define;
  var modules = Object.create(null);

  function define(factory, moduleId, dependencyMap) {
    if (moduleId in modules) {
      return;
    }

    modules[moduleId] = {
      dependencyMap: dependencyMap,
      exports: undefined,
      factory: factory,
      hasError: false,
      isInitialized: false
    };
  }

  function _require(moduleId) {
    var moduleIdReallyIsNumber = moduleId;
    var module = modules[moduleIdReallyIsNumber];
    return module && module.isInitialized ? module.exports : guardedLoadModule(moduleIdReallyIsNumber, module);
  }

  var inGuard = false;

  function guardedLoadModule(moduleId, module) {
    if (!inGuard && global.ErrorUtils) {
      inGuard = true;
      var returnValue = void 0;

      try {
        returnValue = loadModuleImplementation(moduleId, module);
      } catch (e) {
        global.ErrorUtils.reportFatalError(e);
      }

      inGuard = false;
      return returnValue;
    } else {
      return loadModuleImplementation(moduleId, module);
    }
  }

  var ID_MASK_SHIFT = 16;
  var LOCAL_ID_MASK = ~0 >>> ID_MASK_SHIFT;

  function unpackModuleId(moduleId) {
    var segmentId = moduleId >>> ID_MASK_SHIFT;
    var localId = moduleId & LOCAL_ID_MASK;
    return {
      segmentId: segmentId,
      localId: localId
    };
  }

  _require.unpackModuleId = unpackModuleId;

  function packModuleId(value) {
    return value.segmentId << ID_MASK_SHIFT + value.localId;
  }

  _require.packModuleId = packModuleId;

  function loadModuleImplementation(moduleId, module) {
    var nativeRequire = global.nativeRequire;

    if (!module && nativeRequire) {
      var _unpackModuleId = unpackModuleId(moduleId),
          _segmentId = _unpackModuleId.segmentId,
          _localId = _unpackModuleId.localId;

      nativeRequire(_localId, _segmentId);
      module = modules[moduleId];
    }

    if (!module) {
      throw unknownModuleError(moduleId);
    }

    if (module.hasError) {
      throw moduleThrewError(moduleId, module.error);
    }

    module.isInitialized = true;
    var exports = module.exports = {};
    var _module = module,
        factory = _module.factory,
        dependencyMap = _module.dependencyMap;

    try {
      var _moduleObject = {
        exports: exports
      };
      factory(global, _require, _moduleObject, exports, dependencyMap);
      {
        module.factory = undefined;
        module.dependencyMap = undefined;
      }
      return module.exports = _moduleObject.exports;
    } catch (e) {
      module.hasError = true;
      module.error = e;
      module.isInitialized = false;
      module.exports = undefined;
      throw e;
    }
  }

  function unknownModuleError(id) {
    var message = 'Requiring unknown module \\"' + id + '\\".';
    return Error(message);
  }

  function moduleThrewError(id, error) {
    var displayName = id;
    return Error('Requiring module \\"' + displayName + '\\", which threw an exception: ' + error);
  }
})(this);
__d(function (global, _require, module, exports, _dependencyMap) {
  'use strict';

  var Bar = _require(_dependencyMap[0]);

  var Foo = _require(_dependencyMap[1]);

  module.exports = {
    Foo: Foo,
    Bar: Bar
  };
},2,[3,4]);
__d(function (global, _require, module, exports, _dependencyMap) {
  'use strict';

  var Foo = _require(_dependencyMap[0]);

  module.exports = {
    type: 'bar',
    foo: Foo.type
  };
},3,[4]);
__d(function (global, _require, module, exports, _dependencyMap) {
  'use strict';

  var asset = _require(_dependencyMap[0]);

  module.exports = {
    type: 'foo',
    asset: asset
  };
},4,[5]);
__d(function (global, _require, module, exports, _dependencyMap) {
  module.exports = _require(_dependencyMap[0]).registerAsset({
    \\"__packager_asset\\": true,
    \\"httpServerLocation\\": \\"/assets\\",
    \\"width\\": 8,
    \\"height\\": 8,
    \\"scales\\": [1],
    \\"hash\\": \\"77d45c1f7fa73c0f6c444a830dc42f67\\",
    \\"name\\": \\"test\\",
    \\"type\\": \\"png\\"
  });
},5,[6]);
__d(function (global, _require, module, exports, _dependencyMap) {
  'use strict';
},6,[]);
require(2);"
`;
